README.md
# 进度条标记

---

videojs播放器进度条标记插件（videojs6）

## 何时使用

- 需要给videojs播放器的进度条打上标记的时候

## 浏览器支持

IE 9+

## 安装

```bash
npm install videojs-progress-marker --save
```

## 运行

```bash
# 默认开启服务器，地址为 ：http://local:8000/

# 能在ie9+下浏览本站，修改代码后自动重新构建，且能在ie10+运行热更新，页面会自动刷新
npm run start

# 构建生产环境静态文件，用于发布文档
npm run site
```

## 代码演示

### 基本

在进度条添加标记

```jsx
import videojs from 'video.js'
import 'video.js/dist/video-js.css'
import "videojs-progress-marker/lib"
import "videojs-progress-marker/lib/style/"

class App extends React.Component {
  componentDidMount () {
    const node = ReactDOM.findDOMNode(this.videoWrap)
    if (!node) {
      return
    }
    const videoJsOptions = {
      controls: true,
      sources: [{
        src: 'http://os71std62.bkt.clouddn.com/ocean.mp4',
        type: 'video/mp4'
      }],
    }
    // react0.14.x data-reactid问题
    const videoEl = document.createElement('video')
    videoEl.className = `video-js`

    node.appendChild(videoEl)
    this.player = videojs(videoEl, {...videoJsOptions}, () => {
      this.addMarker()
    })
  }
  componentWillUnmount () {
    if (this.player) {
      this.player.dispose()
    }
  }
  addMarker () {
    this.player.markers({
      markers: [
          {time: 9.5, text: "this"},
          {time: 16,  text: "is"},
          {time: 23.6,text: "so"},
          {time: 28,  text: "cool"}
      ]
    })
  }
  render() {
    return (
       <div data-vjs-player ref={node => { this.videoWrap = node }} />
    )
  }
}

ReactDOM.render(<App />, mountNode);
```

### 自定义标记

自定义标记，tootip和overlay的样式和文字

```jsx
import videojs from 'video.js'
import 'video.js/dist/video-js.css'
import "videojs-progress-marker/lib"
import "videojs-progress-marker/lib/style/"

class App extends React.Component {
  componentDidMount () {
    const node = ReactDOM.findDOMNode(this.videoWrap)
    if (!node) {
      return
    }
    const videoJsOptions = {
      controls: true,
      sources: [{
        src: 'http://os71std62.bkt.clouddn.com/ocean.mp4',
        type: 'video/mp4'
      }],
    }
    // react0.14.x data-reactid问题
    const videoEl = document.createElement('video')
    videoEl.className = `video-js`

    node.appendChild(videoEl)
    this.player = videojs(videoEl, {...videoJsOptions}, () => {
      this.addMarker()
    })
  }
  componentWillUnmount () {
    if (this.player) {
      this.player.dispose()
    }
  }
  addMarker () {
    this.player.markers({
      markerStyle: {
      'width':'9px',
      'border-radius': '40%',
      'background-color': 'orange'
      },
      markerTip:{
          display: true,
          text: function(marker) {
            return "I am a marker tip: "+ marker.text;
          }
      },
      breakOverlay:{
          display: true,
          displayTime: 4,
          style:{
            'width':'100%',
            'height': '30%',
            'background-color': 'rgba(10,10,10,0.6)',
            'color': 'white',
            'font-size': '16px'
          },
          text: function(marker) {
            return "This is a break overlay: " + marker.overlayText;
          },
      },
      markers: [
          {time: 9.5, text: "this", overlayText: "1", class: "special-blue"},
          {time: 16,  text: "is", overlayText: "2"},
          {time: 23.6,text: "so", overlayText: "3"},
          {time: 28,  text: "cool", overlayText: "4"}
      ]
    })
  }
  render() {
    return (
       <div data-vjs-player ref={node => { this.videoWrap = node }} />
    )
  }
}

ReactDOM.render(<App />, mountNode);
```

### 动态添加，删除标记

动态添加，删除标记，跳转到标记点

```css
.btn-group button{
  border: 1px  solid #ccc;
  background: #fff;
  padding: 0 10px;
  margin: 10px;
}
```

```jsx
import videojs from 'video.js'
import 'video.js/dist/video-js.css'
import "videojs-progress-marker/lib"
import "videojs-progress-marker/lib/style/"

class App extends React.Component {
  componentDidMount () {
    const node = ReactDOM.findDOMNode(this.videoWrap)
    if (!node) {
      return
    }
    const videoJsOptions = {
      controls: true,
      sources: [{
        src: 'http://os71std62.bkt.clouddn.com/ocean.mp4',
        type: 'video/mp4'
      }],
    }
    // react0.14.x data-reactid问题
    const videoEl = document.createElement('video')
    videoEl.className = `video-js`

    node.appendChild(videoEl)
    this.player = videojs(videoEl, {...videoJsOptions}, () => {
      this.addMarker()
    })
  }
  componentWillUnmount () {
    if (this.player) {
      this.player.dispose()
    }
  }
  addMarker () {
    this.player.markers({
      breakOverlay:{
        display: true
      },
      onMarkerClick: function(marker){
        console.log(`Marker click:${marker.time}`)
      },
      onMarkerReached: function(marker){
        console.log(`Marker reached:${marker.time}`)
      },
      markers: [
          {time: 9.5, text: "this", overlayText: "1"},
          {time: 16,  text: "is", overlayText: "2"},
          {time: 23.6,text: "so", overlayText: "3"},
          {time: 28,  text: "cool", overlayText: "4"}
      ]
    })
  }
  handlePrev = () => {
    this.player.markers.prev()
  }
  handleNext = () => {
    this.player.markers.next()
  }
  handleAdd = () => {
    const randomTime = Math.floor((Math.random() * parseInt(this.player.duration())) + 1);

    // come up with a random time
    this.player.markers.add([{
      time: randomTime,
      text: "I'm new",
      overlayText: "I'm new"
    }]);
  }
  handleMoveForward = () => {
    const markers = this.player.markers.getMarkers();
    for(var i = 0; i < markers.length; i++) {
      markers[i].time += 1;
    }
    this.player.markers.updateTime();
  }
  handleRemoveFirst = () => {
    this.player.markers.remove([0]);
  }
  handleRemoveAll = () => {
    this.player.markers.removeAll();
  }
  handleDestroy = () => {
    this.player.markers.destroy();
  }
  render() {
    return (
      <div> 
        <div data-vjs-player ref={node => { this.videoWrap = node }} />
        <div className="btn-group">
          <button onClick={this.handlePrev}>上一个</button>
          <button onClick={this.handleNext}>下一个</button>
          <button onClick={this.handleAdd}>添加</button>  
          <button onClick={this.handleMoveForward}>向前移动一秒</button>  
          <button onClick={this.handleRemoveFirst}>移除第一个</button>  
          <button onClick={this.handleRemoveAll}>移除所有</button>  
          <button onClick={this.handleDestroy}>销毁</button>  
        </div> 
      </div>
    )
  }
}

ReactDOM.render(<App />, mountNode);
```

## API

代码是fork了[videojs-markers](https://github.com/spchuang/videojs-markers)稍加修改后得来的（videojs-marker没有发布npm包），
API与之相比没有变化，可以查看[这里](http://www.sampingchuang.com/videojs-markers)。


CHANGELOG.md
# 更新日志

#### 发布周期

* 修订版本号：每周末会进行日常 bugfix 更新。（如果有紧急的 bugfix，则任何时候都可发布）
* 次版本号：发布一个带有新特性的向下兼容的版本。
* 主版本号：含有破坏性更新和新特性，不在发布周期内。

---

## 1.0.0

`2018-06-22`

* 修复

## 1.0.0

`2018-08-28`

* 发布第一个正式版本


lint.js
#!/bin/env node
const path = require('path')
const execSync = require('child_process').execSync

const root = path.resolve(__dirname)
// 如果src不在git根目录上，修改成项目需要 lint 的目录
const src = 'src'
const fix = process.env.npm_config_fix ? '--fix' : ''

let error
const exec = function (cmd) {
  try {
    return execSync(cmd, { stdio: [process.stdin, process.stdout, process.stderr] })
  } catch (e) {
    error = true
  }
}

exec(`stylelint "${root}/${src}/**/*.css" ${fix}`)
exec(`stylelint "${root}/${src}/**/*.scss" ${fix}`)
exec(`stylelint "${root}/${src}/**/*.less" ${fix}`)
exec(`eslint ${fix} --ext .js,.jsx ${root}/${src} `)

if (error) {
  if (fix) {
    throw new Error('部分错误已经自动修正，剩余的请手动进行处理。可以再次运行 npm run lint 进行检查')
  } else {
    throw new Error('可以尝试运行 npm run lint --fix 进行错误自动修复')
  }
} else {
  console.log('完美！代码规范检查没有发现任何错误信息!')
}


src/index.js
import videojs from 'video.js'

// default setting
const defaultSetting = {
  markerStyle: {
    'width': '7px',
    'border-radius': '30%',
    'background-color': 'red'
  },
  markerTip: {
    display: true,
    text: function (marker) {
      return 'Break: ' + marker.text
    },
    time: function (marker) {
      return marker.time
    }
  },
  breakOverlay: {
    display: false,
    displayTime: 3,
    text: function (marker) {
      return 'Break overlay: ' + marker.overlayText
    },
    style: {
      'width': '100%',
      'height': '20%',
      'background-color': 'rgba(0,0,0,0.7)',
      'color': 'white',
      'font-size': '17px'
    }
  },
  onMarkerClick: function (marker) {},
  onMarkerReached: function (marker, index) {},
  markers: []
}

// create a non-colliding random number
function generateUUID () {
  let d = new Date().getTime()
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (d + Math.random() * 16) % 16 | 0
    d = Math.floor(d / 16)
    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16)
  })
}

/**
 * Returns the size of an element and its position
 * a default Object with 0 on each of its properties
 * its return in case there's an error
 * @param  {Element} element  el to get the size and position
 * @return {DOMRect|Object}   size and position of an element
 */
function getElementBounding (element) {
  let elementBounding
  const defaultBoundingRect = {
    top: 0,
    bottom: 0,
    left: 0,
    width: 0,
    height: 0,
    right: 0
  }

  try {
    elementBounding = element.getBoundingClientRect()
  } catch (e) {
    elementBounding = defaultBoundingRect
  }

  return elementBounding
}

const NULL_INDEX = -1

function registerVideoJsMarkersPlugin (options) {
  // copied from video.js/src/js/utils/merge-options.js since
  // videojs 4 doens't support it by defualt.
  if (!videojs.mergeOptions) {
    function isPlain (value) {
      return !!value && typeof value === 'object' &&
        toString.call(value) === '[object Object]' &&
        value.constructor === Object
    }
    function mergeOptions (source1, source2) {
      const result = {}
      const sources = [source1, source2]
      sources.forEach(source => {
        if (!source) {
          return
        }
        Object.keys(source).forEach(key => {
          const value = source[key]
          if (!isPlain(value)) {
            result[key] = value
            return
          }
          if (!isPlain(result[key])) {
            result[key] = {}
          }
          result[key] = mergeOptions(result[key], value)
        })
      })
      return result
    }
    videojs.mergeOptions = mergeOptions
  }

  if (!videojs.dom.createEl) {
    videojs.dom.createEl = function (tagName, props, attrs) {
      const el = videojs.Player.prototype.dom.createEl(tagName, props)
      if (!!attrs) {
        Object.keys(attrs).forEach(key => {
          el.setAttribute(key, attrs[key])
        })
      }
      return el
    }
  }

  /**
   * register the markers plugin (dependent on jquery)
   */
  const setting = videojs.mergeOptions(defaultSetting, options)
  const markersMap = {}
  const markersList = [] // list of markers sorted by time
  let currentMarkerIndex = NULL_INDEX
  const player = this
  let markerTip = null
  let breakOverlay = null
  let overlayIndex = NULL_INDEX

  function sortMarkersList () {
    // sort the list by time in asc order
    markersList.sort((a, b) => {
      return setting.markerTip.time(a) - setting.markerTip.time(b)
    })
  }

  function addMarkers (newMarkers) {
    newMarkers.forEach((marker) => {
      marker.key = generateUUID()

      player.el().querySelector('.vjs-progress-holder')
        .appendChild(createMarkerDiv(marker))

      // store marker in an internal hash map
      markersMap[marker.key] = marker
      markersList.push(marker)
    })

    sortMarkersList()
  }

  function getPosition (marker) {
    return (setting.markerTip.time(marker) / player.duration()) * 100
  }

  function setMarkderDivStyle (marker, markerDiv) {
    markerDiv.className = `vjs-marker ${marker.class || ''}`

    Object.keys(setting.markerStyle).forEach(key => {
      markerDiv.style[key] = setting.markerStyle[key]
    })

    // hide out-of-bound markers
    const ratio = marker.time / player.duration()
    if (ratio < 0 || ratio > 1) {
      markerDiv.style.display = 'none'
    }

    // set position
    markerDiv.style.left = getPosition(marker) + '%'
    if (marker.duration) {
      markerDiv.style.width = (marker.duration / player.duration()) * 100 + '%'
      markerDiv.style.marginLeft = '0px'
    } else {
      const markerDivBounding = getElementBounding(markerDiv)
      markerDiv.style.marginLeft = markerDivBounding.width / 2 + 'px'
    }
  }

  function createMarkerDiv (marker) {
    const markerDiv = videojs.dom.createEl('div', {}, {
      'data-marker-key': marker.key,
      'data-marker-time': setting.markerTip.time(marker)
    })

    setMarkderDivStyle(marker, markerDiv)

    // bind click event to seek to marker time
    markerDiv.addEventListener('click', function (e) {
      let preventDefault = false
      if (typeof setting.onMarkerClick === 'function') {
        // if return false, prevent default behavior
        preventDefault = setting.onMarkerClick(marker) === false
      }

      if (!preventDefault) {
        const key = this.getAttribute('data-marker-key')
        player.currentTime(setting.markerTip.time(markersMap[key]))
      }
    })

    if (setting.markerTip.display) {
      registerMarkerTipHandler(markerDiv)
    }

    return markerDiv
  }

  function updateMarkers (force) {
    // update UI for markers whose time changed
    markersList.forEach((marker) => {
      const markerDiv = player.el().querySelector(".vjs-marker[data-marker-key='" + marker.key + "']")
      const markerTime = setting.markerTip.time(marker)

      if (force || markerDiv.getAttribute('data-marker-time') !== markerTime) {
        setMarkderDivStyle(marker, markerDiv)
        markerDiv.setAttribute('data-marker-time', markerTime)
      }
    })
    sortMarkersList()
  }

  function removeMarkers (indexArray) {
    // reset overlay
    if (!!breakOverlay) {
      overlayIndex = NULL_INDEX
      breakOverlay.style.visibility = 'hidden'
    }
    currentMarkerIndex = NULL_INDEX

    const deleteIndexList = []
    indexArray.forEach((index) => {
      const marker = markersList[index]
      if (marker) {
        // delete from memory
        delete markersMap[marker.key]
        deleteIndexList.push(index)

        // delete from dom
        const el = player.el().querySelector(".vjs-marker[data-marker-key='" + marker.key + "']")
        el && el.parentNode.removeChild(el)
      }
    })

    // clean up markers array
    deleteIndexList.reverse()
    deleteIndexList.forEach((deleteIndex) => {
      markersList.splice(deleteIndex, 1)
    })

    // sort again
    sortMarkersList()
  }

  // attach hover event handler
  function registerMarkerTipHandler (markerDiv) {
    markerDiv.addEventListener('mouseover', () => {
      const marker = markersMap[markerDiv.getAttribute('data-marker-key')]
      if (!!markerTip) {
        markerTip.querySelector('.vjs-tip-inner').innerText = setting.markerTip.text(marker)
        // margin-left needs to minus the padding length to align correctly with the marker
        markerTip.style.left = getPosition(marker) + '%'
        const markerTipBounding = getElementBounding(markerTip)
        const markerDivBounding = getElementBounding(markerDiv)
        markerTip.style.marginLeft =
          -parseFloat(markerTipBounding.width / 2) + parseFloat(markerDivBounding.width / 4) + 'px'
        markerTip.style.visibility = 'visible'
      }
    })

    markerDiv.addEventListener('mouseout', () => {
      if (!!markerTip) {
        markerTip.style.visibility = 'hidden'
      }
    })
  }

  function initializeMarkerTip () {
    markerTip = videojs.dom.createEl('div', {
      className: 'vjs-tip',
      innerHTML: "<div class='vjs-tip-arrow'></div><div class='vjs-tip-inner'></div>"
    })
    player.el().querySelector('.vjs-progress-holder').appendChild(markerTip)
  }

  // show or hide break overlays
  function updateBreakOverlay () {
    if (!setting.breakOverlay.display || currentMarkerIndex < 0) {
      return
    }

    const currentTime = player.currentTime()
    const marker = markersList[currentMarkerIndex]
    const markerTime = setting.markerTip.time(marker)

    if (
      currentTime >= markerTime &&
      currentTime <= (markerTime + setting.breakOverlay.displayTime)
    ) {
      if (overlayIndex !== currentMarkerIndex) {
        overlayIndex = currentMarkerIndex
        if (breakOverlay) {
          breakOverlay.querySelector('.vjs-break-overlay-text').innerHTML = setting.breakOverlay.text(marker)
        }
      }

      if (breakOverlay) {
        breakOverlay.style.visibility = 'visible'
      }
    } else {
      overlayIndex = NULL_INDEX
      if (breakOverlay) {
        breakOverlay.style.visibility = 'hidden'
      }
    }
  }

  // problem when the next marker is within the overlay display time from the previous marker
  function initializeOverlay () {
    breakOverlay = videojs.dom.createEl('div', {
      className: 'vjs-break-overlay',
      innerHTML: "<div class='vjs-break-overlay-text'></div>"
    })
    Object.keys(setting.breakOverlay.style).forEach(key => {
      if (breakOverlay) {
        breakOverlay.style[key] = setting.breakOverlay.style[key]
      }
    })
    player.el().appendChild(breakOverlay)
    overlayIndex = NULL_INDEX
  }

  function onTimeUpdate () {
    onUpdateMarker()
    updateBreakOverlay()
    options.onTimeUpdateAfterMarkerUpdate && options.onTimeUpdateAfterMarkerUpdate()
  }

  function onUpdateMarker () {
    /*
      check marker reached in between markers
      the logic here is that it triggers a new marker reached event only if the player
      enters a new marker range (e.g. from marker 1 to marker 2).
      Thus, if player is on marker 1 and user clicked on marker 1 again, no new reached event is triggered)
    */
    if (!markersList.length) {
      return
    }

    const getNextMarkerTime = (index) => {
      if (index < markersList.length - 1) {
        return setting.markerTip.time(markersList[index + 1])
      }
      // next marker time of last marker would be end of video time
      return player.duration()
    }
    const currentTime = player.currentTime()
    let newMarkerIndex = NULL_INDEX

    if (currentMarkerIndex !== NULL_INDEX) {
      // check if staying at same marker
      const nextMarkerTime = getNextMarkerTime(currentMarkerIndex)
      if (
        currentTime >= setting.markerTip.time(markersList[currentMarkerIndex]) &&
        currentTime < nextMarkerTime
      ) {
        return
      }

      // check for ending (at the end current time equals player duration)
      if (
        currentMarkerIndex === markersList.length - 1 &&
        currentTime === player.duration()
      ) {
        return
      }
    }

    // check first marker, no marker is selected
    if (currentTime < setting.markerTip.time(markersList[0])) {
      newMarkerIndex = NULL_INDEX
    } else {
      // look for new index
      for (let i = 0; i < markersList.length; i++) {
        const nextMarkerTime = getNextMarkerTime(i)
        if (
          currentTime >= setting.markerTip.time(markersList[i]) &&
          currentTime < nextMarkerTime
        ) {
          newMarkerIndex = i
          break
        }
      }
    }

    // set new marker index
    if (newMarkerIndex !== currentMarkerIndex) {
      // trigger event if index is not null
      if (newMarkerIndex !== NULL_INDEX && options.onMarkerReached) {
        options.onMarkerReached(markersList[newMarkerIndex], newMarkerIndex)
      }
      currentMarkerIndex = newMarkerIndex
    }
  }

  // setup the whole thing
  function initialize () {
    if (setting.markerTip.display) {
      initializeMarkerTip()
    }

    // remove existing markers if already initialized
    player.markers.removeAll()
    addMarkers(setting.markers)

    if (setting.breakOverlay.display) {
      initializeOverlay()
    }
    onTimeUpdate()
    player.on('timeupdate', onTimeUpdate)
    player.off('loadedmetadata')
  }

  // setup the plugin after we loaded video's meta data
  player.on('loadedmetadata', function () {
    initialize()
  })

  // exposed plugin API
  player.markers = {
    getMarkers: function () {
      return markersList
    },
    next: function () {
      // go to the next marker from current timestamp
      const currentTime = player.currentTime()
      for (let i = 0; i < markersList.length; i++) {
        const markerTime = setting.markerTip.time(markersList[i])
        if (markerTime > currentTime) {
          player.currentTime(markerTime)
          break
        }
      }
    },
    prev: function () {
      // go to previous marker
      const currentTime = player.currentTime()
      for (let i = markersList.length - 1; i >= 0; i--) {
        const markerTime = setting.markerTip.time(markersList[i])
        // add a threshold
        if (markerTime + 0.5 < currentTime) {
          player.currentTime(markerTime)
          return
        }
      }
    },
    add: function (newMarkers) {
      // add new markers given an array of index
      addMarkers(newMarkers)
    },
    remove: function (indexArray) {
      // remove markers given an array of index
      removeMarkers(indexArray)
    },
    removeAll: function () {
      const indexArray = []
      for (let i = 0; i < markersList.length; i++) {
        indexArray.push(i)
      }
      removeMarkers(indexArray)
    },
    // force - force all markers to be updated, regardless of if they have changed or not.
    updateTime: function (force) {
      // notify the plugin to update the UI for changes in marker times
      updateMarkers(force)
    },
    reset: function (newMarkers) {
      // remove all the existing markers and add new ones
      player.markers.removeAll()
      addMarkers(newMarkers)
    },
    destroy: function () {
      // unregister the plugins and clean up even handlers
      player.markers.removeAll()
      breakOverlay && breakOverlay.remove()
      markerTip && markerTip.remove()
      player.off('timeupdate', updateBreakOverlay)
      delete player.markers
    }
  }
}

videojs.registerPlugin('markers', registerVideoJsMarkersPlugin)
